runtime/near-vm-logic/src/vmstate.rs

```diff
@@ -0,0 +1,177 @@
+        offset: u64,
+        buf: &[u8],
+    ) -> Result<()> {
+        gas_counter.pay_base(write_memory_base)?;
+        gas_counter.pay_per(write_memory_byte, buf.len() as _)?;
+        self.0.write_memory(offset, buf).map_err(|_| HostError::MemoryAccessViolation.into())
+    }
+
+    memory_get!(u128, get_u128);
+    memory_get!(u32, get_u32);
+    memory_get!(u16, get_u16);
+    memory_get!(u8, get_u8);
+    memory_set!(u128, set_u128);
+}
+
+/// Registers to use by the guest.
+///
+/// Provides interface to access registers while correctly accounting for gas
+/// usage.
+///
+/// See documentation of [`Memory`] for more motivation for this struct.
+#[derive(Default, Clone)]
+pub(super) struct Registers(std::collections::HashMap<u64, Box<[u8]>>);
+
+impl Registers {
+    /// Returns register with given index.
+    ///
+    /// Returns an error if (i) there’s not enough gas to perform the register
+    /// read or (ii) register with given index doesn’t exist.
+    pub(super) fn get<'s>(
+        &'s self,
+        gas_counter: &mut GasCounter,
+        register_id: u64,
+    ) -> Result<&'s [u8]> {
+        if let Some(data) = self.0.get(&register_id) {
+            gas_counter.pay_base(read_register_base)?;
+            gas_counter.pay_per(read_register_byte, data.len() as u64)?;
+            Ok(&data[..])
+        } else {
+            Err(HostError::InvalidRegisterId { register_id }.into())
+        }
+    }
+
+    /// Returns length of register with given index or None if no such register.
+    pub(super) fn get_len(&self, register_id: u64) -> Option<u64> {
+        self.0.get(&register_id).map(|data| data.len() as u64)
+    }
+
+    /// Sets register with given index.
+    ///
+    /// Returns an error if (i) there’s not enough gas to perform the register
+    /// read or (ii) if setting the register would violate configured limits.
+    pub(super) fn set(
+        &mut self,
+        gas_counter: &mut GasCounter,
+        config: &VMLimitConfig,
+        register_id: u64,
+        data: impl Into<Box<[u8]>>,
+    ) -> Result<()> {
+        let data = data.into();
+        let data_len = data.len() as u64;
+        gas_counter.pay_base(write_register_base)?;
+        gas_counter.pay_per(write_register_byte, data_len)?;
+        // Fun fact: if we are at the limit and we replace a register, we’ll
+        // fail even though we should be succeeding.  This bug is now part of
+        // the protocol so we can’t change it.
+        if data_len > config.max_register_size || self.0.len() as u64 >= config.max_number_registers
+        {
+            return Err(HostError::MemoryAccessViolation.into());
+        }
+        match self.0.insert(register_id, data) {
+            Some(old_value) if old_value.len() as u64 >= data_len => {
+                // If there was old value and it was no shorter than the new
+                // one, there’s no need to check new memory usage since it
+                // didn’t increase.
```

Comment:
This is another gotcha for the fees estimation – if our estimation benchmark ends up hitting this branch more often than not, we’d fail to price in the `O(num_registers)` computation cost that the other branch incurs…

Reply:
This should be solved if this is changed to remember the usage in a field.